pkg/policy/mapstate.go

```diff
@@ -498,24 +629,198 @@ func (keys MapState) denyPreferredInsertWithChanges(newKey Key, newEntry MapStat
 				}
 			}
 		}
-		keys.redirectPreferredInsert(newKey, newEntry, adds, deletes)
+		keys.authPreferredInsert(newKey, newEntry, features, changes)
 	}
 }
 
-// redirectPreferredInsert inserts a new entry giving priority to L7-redirects by
-// not overwriting a L7-redirect entry with a non-redirect entry.
-func (keys MapState) redirectPreferredInsert(key Key, entry MapStateEntry, adds, deletes Keys) {
-	// Do not overwrite the entry, but only merge owners if the old entry is a deny or redirect.
-	// This prevents an existing deny or redirect being overridden by a non-deny or a non-redirect.
-	// Merging owners from the new entry to the existing one has no datapath impact so we skip
-	// adding anything to 'adds' here.
-	if oldEntry, exists := keys[key]; exists && (oldEntry.IsRedirectEntry() || oldEntry.IsDeny) {
-		oldEntry.MergeReferences(&entry)
-		keys[key] = oldEntry
-		return
+// IsSuperSetOf checks if the receiver Key is a superset of the argument Key, and returns a
+// specificity score of the receiver key (higher score is more specific), if so. Being a superset
+// means that the receiver key would match all the traffic of the argument key without being the
+// same key. Hence, a L3-only key is not a superset of a L4-only key, as the L3-only key would match
+// the traffic for the given L3 only, while the L4-only key matches traffic on the given port for
+// all the L3's.
+// Returns 0 if the receiver key is not a superset of the argument key.
+//
+// Specificity score for all possible superset wildcard patterns. Datapath requires proto to be specified if port is specified.
+// x. L3/proto/port
+//  1. */*/*
+//  2. */proto/*
+//  3. */proto/port
+//  4. ID/*/*
+//  5. ID/proto/*
+//     ( ID/proto/port can not be superset of anything )
+func (k Key) IsSuperSetOf(other Key) int {
+	if k.TrafficDirection != other.TrafficDirection {
+		return 0 // TrafficDirection must match for 'k' to be a superset of 'other'
+	}
+	if k.Identity == 0 {
```

Comment:
Do we need to think about CIDRS in this method?
```suggestion
	if k.Identity == 0 ||
		ip.NetsContainsAny(keyEntry.getNets(identities, key.Identity),
			otherEntry.getNets(identities, other.Identity)) {
```
Obviously the method would need to be refactored to accept the entries, and also the `identities` interface. I'm fairly certain the specificity math does not need to change even if this gets added.

Reply:
My gut feeling is that CIDRs don't need to be considered, because prior to policy evaluation the ipcache will already apply LPM to identify the most specific identity for the policymap entry. Given that this identity is label-selected in the same way as any other policy evaluation logic, for more specific CIDRs there will be additional identities and hence additional map entries to cover each superset or subset CIDR. The only conflicts we need to deal with here are interactions between wildcard entries (eg "applies to all L3s" or "applies to all L4s") and more specific identity policies (where some of the L3/L4 fields are the same, while the others are wildcarded). 

Put another way, CIDR identity ambiguity is already handled by the LPM ipcache lookup earlier in the datapath, so this means that the policy engine doesn't have to account for it; the policy posture for each explicit CIDR is individually evaluated.

That said, if you have a particular set of overlapping policy rules in mind to consider for this case, that would be super helpful both to reason through and also to ensure the tests are covering that case.

Reply:
Joe and I discussed this in person and decided that CIDRs aren't relevant to Auth. That is to say, no CIDR policy will currently select an endpoint that Cilium can enforce authentication or authorization on.

Reply:
Related, we dug fairly deep into some of the reasons why the deny policy resolution logic needs to take into account CIDR identity supersets, and I've tried to document some of the background in this PR: https://github.com/cilium/cilium/pull/26720 . This may also be a helpful reference to contrast with the auth policy cases to see how CIDR identity supersets may or may not impact this policy evaluation logic in future.